ARCs/arc-0052.md

```diff
@@ -0,0 +1,210 @@
+---
+arc: 52
+title: Wallet Contextual KeyGen and Signing
+description: Algorand Wallet API for Contextual KeyGen and Signing
+author: Bruno Martins (@ehanoc), Patrick Bennett (@pbennett), Kevin Wellenzohn (@k13n)
+discussions-to: https://github.com/algorandfoundation/ARCs/issues/239
+status: Draft
+type: Standards Track
+category: Interface
+created: 2023-06-30
+---
+
+# Algorand Wallet Contextual KeyGen and Signing
+
+## Abstract
+
+This document specifies an expansion of the Algorand Wallet API to support contextual key generation and signing for **non-algorand-transaction** contexts.
+
+## Specification
+
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.
+
+> Comments like this are non-normative.
+
+### Overview
+
+> This overview section is non-normative.
+
+At a high level this specification defines a new API to generate keys and signing in other contexts than transactions.
+
+> Interfaces are defined in TypeScript. All the objects that are defined are valid JSON objects.
+
+#### Contexts
+
+An enum of contexts that are supported by the wallet. This enum is extensible and can be expanded in the future to support more contexts (i.e messaging, ephemeral keys, other chain derivations, etc)
+
+```ts
+enum KeyContext {
+    Address = 0, // Otherwise known as Account?
+    Identity = 1,
+    ...
+}
+```
+
+#### BIP44 Context Derivation Paths
+
+There is several reasons why we want to avoid key re-use in different contexts. One of the reasons is that we want to avoid the possibility of **leaking** information about the identity of the user. Another reason is that there is some cryptographic security concers when re-using keys or in case of keys being compromised the impact is **compartmentalized**.
+
+For this reason is advisable to have different derivation paths for different contexts supporting recoveribility of the keys by using the same mnemonic as before.
+
+| Purpose | Coin Type | Account | Change | Address Index |    Context    |
+| :-----: | :-------: | :-----: | :----: | :-----------: | :-----------: |
+|   44'   |   283'    |   x'    |   x    |       x       |   Addresses   |
+|   44'   |    0'     |   x'    |   x    |       x       | Identity Keys |
+
+see <a href="https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki"> `BIP-44` </a>
+
+#### Wallet New Interface
+
+This new interface is for the wallet to implement. It is an extension of the current wallet interface.
+
+This extension is **OPTIONAL** and separate from the current interface that handles address generation and transaction signing.
+
+```ts
+interface Wallet {
+    ...
+    keyGen(context: KeyContext, account:number, keyIndex: number): Promise<Uint8Array>;
```

Comment:
I am ok with having those well specified in the implementation. However, I would like to be mandatory in the spec as well.

IMHO those constraints should be clarified in the ARC-52 spec, or at least should be redirected to BIP-32. My issue with the current text is that an Algorand wallet might be compliant with ARC-52 but not compliant with BIP-32.

Reply:
BIP-44 defines the hardening at account / index level. 

I don't think we should expose that here as this is a wallet API. Anyone that would like to do BIP32 but not BIP44 compliant, should just use a low level BIP32 API to derive keys that way (i.e non hardened accounts or hardened indexes instead)

Reply:
I understand, although i see problems with that. 

We could expose more options or a more "advanced" API. problem is not every integrator understands the nuances of technical standards that are used underneath. Good API's in my opinion are strict and don't offer too many options that can be miss used. 

There can be an integrator that understands BIP-32 and another that doesn't. This would lead for implementations that might not be interoperable and recovery in one wallet wouldn't work on another wallet.